### ISSUE TITLE
Delay log density computation in pyro.factor

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/pyro#2744]
In several places (e.g. [NeuTra reparam](https://github.com/pyro-ppl/pyro/blob/7a69a3cf8b62f398607b6180dda978029cde698c/pyro/infer/reparam/neutra.py#L88)) that we used `pyro.factor` or `dist.Delta`, log_density is computed explicitly. This computation is not needed, e.g. when making predictions. It would be nice to add a mechanism to defer this computation, only until some inference algorithms that need the log probability.

### Proposed solution

Allow [log_factor](http://docs.pyro.ai/en/stable/primitives.html#pyro.primitives.factor) to be a callable in `pyro.factor`. In `dist.Unit.log_prob`, if `log_factor` is a callable, we will call it to get the required value (similar to the behavior in `pyro.param`).

**Comment by fritzo** [comment]
Here's a possible solution that avoids the need for non-tensors in distributions (which would complicate typing):

What if we use `poutine.mask(mask=False)` to control whether various computations are needed? During prediction we can install a `mask(False)` handler to disable computation of log density. A brute force way to do this would be to add a primitive `is_mask_false()` or something. Or instead of a new `poutine`, we could in `NeuTra` add logic like this:

<details>

```diff
- z_unconstrained = pyro.sample("{}_shared_latent".format(name),
-                               self.guide.get_base_dist().mask(False))
+ with poutine.trace() as trace:
+     z_name = f"{name}_shared_latent"
+     z_unconstrained = pyro.sample(z_name, self.guide.get_base_dist().mask(False))
+ z_mask = trace.nodes["z_name"]["mask"]
  
  # Differentiably transform.
  x_unconstrained = self.transform(z_unconstrained)
- log_density = self.transform.log_abs_det_jacobian(z_unconstrained, x_unconstrained)
+ if z_mask is False:
+     log_density = z_unconstrained.new_zeros(())
+ else:
+     log_density = self.transform.log_abs_det_jacobian(z_unconstrained, x_unconstrained)
  self.x_unconstrained = list(reversed(list(self.guide._unpack_latent(x_unconstrained))))
```

</details>

**Comment by fehiepsi** [comment]
Thanks, @fritzo! Let's discuss more details in the meeting. I think something like `is_mask_false()` will work though still prefer the lambda solution.

**Comment by fritzo** [comment]
Hi @fehiepsi, to aid discussion I've implemented one of our two ideas in #2751. However I'd like to better understand your use case, in particular, when would you draw samples from NeuTra but not need the log density? Is that usage limited to inside `Predictive` where we can wrap in `torch.no_grad()` and `poutine.mask(mask=False)`?

I guess I see the pros & cons of the two approaches as:

1. Allowing `log_factor` to be a callable in `pyro.factor`, `Unit`, and `Delta`:
    - pro: avoids need for calling code to wrap in `poutine.mask` or similar; optimization is automatic
    - con: requires all inference code to create a local callback function
    - con: `Unit` and `Delta` distributions may not know their batch shape, and cannot be validated
      (but we could easily default to eagerly evaluating the callable if `is_validation_enabled()`)
2. Throttling with `poutine.mask(mask=False)`:
    - pro: `Unit` and `Delta` distributions still have tensors, still have shape, and can still be validated
    - con: requires all inference code to consult `poutine.inspect()["mask"] is False`
    - con: requires all calling code to wrap in `poutine.mask(mask=False)`

**Comment by fehiepsi** [comment]
Hi @fritzo, the original motivation is for NeuTra, where latent samples will become deterministic sites, which are transformed by NeuTra from base latent samples. So after running MCMC to get base latent samples, we need to do another pass to get those deterministic values. The transform log_det is a bit expensive to calculate.

Recently, we have a new HMCECS sampler in NumPyro. To utilize it effectively, we need control variates to bias-corrected the likelihood (not log-likelihood). The current way to do it is to mask out the original likelihood, recalculate it using covariates, and add a factor term. Calculating the covariates and factors is a bit expensive (depending on subsample_size) so it would be nice to avoid it when predicting.

This optimization also applies for our current autoguide, where computing transforms' logdet is not necessary in prediction.

Thanks for making `inspect` handler! It should solve the issue, though I don't think we need to expose this handler. Other option is to have something similar to `pyro.factor`, but requires both `shape` and a callable. Both ways require similar amount of internal code I guess. For users, a `factor` solution might be more convenient.

**Comment by fritzo** [comment]
Looking at more real-world models, I see there is often other code surrounding `pyro.factor`, including plates and poutines:
```py
def model():
    ...
    # This whole section could be omitted during prediction:
    tree_scale = pyro.sample("tree_scale", dist.LogNormal(0, 5))
    with pyro.plate("edges", len(edges), dim=-1):
        u, v = log_rate[edges].unbind(-1)
        pyro.sample(
            "rate_change",
            dist.Laplace(0, tree_scale),
            obs=u - v,
        )
    ...
```
One advantage of a check like `is_masked()` is that we could use it around entire sections of model code. In particular plates and poutines would not be able to be embedded inside a `lambda` passed to `pyro.factor`.
